### DISCUSSION TITLE
Any examples showing how to use parallel search without using Script-based objects?

### DISCUSSION THREAD

**Discussion by yurivict** [Gecode/gecode#121] [General]
The example ```golomb-ruler.cpp``` uses ```IntMinimizeScript::run()``` with the ```threads``` argument passed through the ```SizeOptions``` object.

But are there examples showing how to use parallel search with bare engine objects? For example:
```
  BAB<Solver> e;
  while (auto s = e.next()) {
     ...
  }
```

**Comment by Dekker1** [comment]
`Script` objects are probably the best way (and maybe the only recommended way) of interacting with the search process in Gecode. It might be worth reconsidering implementing your goals as part of a custom `Script` class instead of trying to re-invent the wheel.

But to point you in the direction of the information: `threads` is an option of the Search Engine, not an option for the Script. You can find a lot of information about how engines work and everything else about search in Chapter 9 (Search) of [Modelling and Programming with Gecode](https://www.gecode.org/doc-latest/MPG.pdf). When working with Gecode this should generally be your first resource.

Just in case you are looking for more examples, there is also no harm in looking at how the script you mentioned is implemented:
- `IntMinimizeScript` is declared [here](https://github.com/Gecode/gecode/blob/027c57889d66dd26ad8e1a419c2cda22ab0cf305/gecode/driver.hh#L807-L808).
- It's actual "run" method can be found [here](https://github.com/Gecode/gecode/blob/027c57889d66dd26ad8e1a419c2cda22ab0cf305/gecode/driver/script.hpp#L363-L456)

**Comment by zayenz** [comment]
As @Dekker1 says, the `threads` is an option that is passed through to the search engine using [Search::Options](https://www.gecode.org/doc-latest/reference/classGecode_1_1Search_1_1Options.html) that can be passed to a search engine to configure it.

Here is a small example on how to use search options to set up a parallel search using 3 threads and additionally letting the search engine take ownership of the root node.
```c++
  Gecode::Search::Options options;
  // Use 3 threads.
  options.threads = 3;
  // Take ownership of the root node passed into the search engine
  options.clone = false;
  auto* root = new Sudoku(instance);
  Gecode::DFS<Sudoku> engine(root, options);
  while (auto* solution = engine.next()) {
    solution->print(std::cout);
    delete solution;
  }
```

The threads argument is documented in [Section 9.3.1 in MPG](https://www.gecode.org/doc-latest/MPG.pdf).

**Comment by yurivict** [comment]
Thanks Mikael, this code snippet is exactly what I was looking for.

-------------------

The reason I avoided ```IntMinimizeScript``` is that it does more than I need. For example, it prints solutions unconditionally.
